# Customer Retention and Churn Analysis

## Task 2 | Future Interns - Data Science & Analytics Internship (2026)

### Project Overview

This project focuses on Customer Retention & Churn Analysis, evaluating subscription-based behavior to identify why customers leave and what drives long-term loyalty. By applying cohort analysis and lifetime metrics, the goal is to provide data-driven strategies that minimize churn and maximize the Customer Lifetime Value (CLV).

### Dataset Description

The analysis is conducted using the Telco Customer Churn Dataset, which contains information about 7,043 customers at a California-based telecommunications company. It includes demographics, account information (tenure, contract type, payment method), and service usage details.

### Objectives & Methodology

The workflow for this project involves:

- Data Preparation: Handling missing values in total charges, encoding categorical variables, and segmenting tenure into logical cohorts.
- Cohort Analysis: Grouping customers by their signup period and contract type to track retention behavior over time.
- Feature Correlation: Identifying the key drivers—such as contract duration or technical support access—that have the highest correlation with customer churn.
- Predictive Insights: Using exploratory data analysis to pinpoint high-risk segments before they churn.
- Strategic Reporting: Developing a Power BI Dashboard that translates complex retention metrics into actionable business recommendations.

### Research Questions

To guide the analysis, this project seeks to answer:

- Churn Patterns: What is the overall churn rate, and how does it fluctuate across different customer segments (e.g., Senior Citizens vs. Families)?
- Retention Drivers: Which services (Online Security, Tech Support, Streaming) are most effective at keeping customers engaged?
- Contractual Impact: How do month-to-month contracts compare to one- or two-year commitments in terms of stability and revenue?
- Revenue Leakage: What is the estimated monthly revenue loss associated with churning customers?
- Actionable Strategy: What specific interventions (e.g., loyalty programs or plan upgrades) can be implemented to improve retention in high-risk groups?

## DATA PREPROCESSING

In [44]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [45]:
data = pd.read_csv('customer_data.csv')

In [46]:
data.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [47]:
data.tail()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
7038,6840-RESVB,Male,0,Yes,Yes,24,Yes,Yes,DSL,Yes,...,Yes,Yes,Yes,Yes,One year,Yes,Mailed check,84.80,1990.5,No
7039,2234-XADUH,Female,0,Yes,Yes,72,Yes,Yes,Fiber optic,No,...,Yes,No,Yes,Yes,One year,Yes,Credit card (automatic),103.20,7362.9,No
7040,4801-JZAZL,Female,0,Yes,Yes,11,No,No phone service,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.60,346.45,No
7041,8361-LTMKD,Male,1,Yes,No,4,Yes,Yes,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Mailed check,74.40,306.6,Yes
7042,3186-AJIEK,Male,0,No,No,66,Yes,No,Fiber optic,Yes,...,Yes,Yes,Yes,Yes,Two year,Yes,Bank transfer (automatic),105.65,6844.5,No


This dataset provides details on customer demographic traits, service subscriptions, and billing details are linked to a binary Churn indicator, providing the foundation for identifying which specific variables most strongly correlate with customer loss.

In [48]:
data.shape

(7043, 21)

The dataset contains 21 column attributes and 7043 row entries.

In [49]:
data.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


The summary statistics indicate a diverse customer base with an average tenure of roughly 32 months and monthly charges averaging 64.76, while the low mean for senior citizens (0.16) suggests the majority of the subscriber population is younger.

## DATA CLEANING

In [50]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


There is a need for data type conversion as TotalCharges is currently stored as an object rather than a numeric float.

In [51]:
data['TotalCharges'] = pd.to_numeric(data['TotalCharges'], errors='coerce')

data['TotalCharges'] = data['TotalCharges'].fillna(0)

In [52]:
data.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

There are no missing values in the data so we go on to check for duplicates

In [53]:
data.duplicated().sum()

0

There are no duplicates in the dataset so we check for outliers

In [54]:
kurtosis_values = data.select_dtypes(include=['number']).kurt()

print("Kurtosis values:")
print(kurtosis_values)

Kurtosis values:
SeniorCitizen     1.362596
tenure           -1.387372
MonthlyCharges   -1.257260
TotalCharges     -0.228580
dtype: float64


Kurtosis values are less than 3 which means we likely do not have extreme values or errors. The kurtosis values confirm that the data is non-normally distributed, specifically highlighting that customers cluster at the extreme ends of the tenure spectrum rather than around a central average. We also see that monthly billing is highly diverse and segmented while the total revenue contributed per customer follows a more predictable, traditional pattern.

In [55]:
# standardise column names 

rename_dict = {
    'customerID': 'customer_id',
    'SeniorCitizen': 'senior_citizen',
    'Partner': 'partner',
    'Dependents': 'dependents',
    'PhoneService': 'phone_service',
    'MultipleLines': 'multiple_lines',
    'InternetService': 'internet_service',
    'OnlineSecurity': 'online_security',
    'OnlineBackup': 'online_backup',
    'DeviceProtection': 'device_protection',
    'TechSupport': 'tech_support',
    'StreamingTV': 'streaming_tv',
    'StreamingMovies': 'streaming_movies',
    'Contract': 'contract',
    'PaperlessBilling': 'paperless_billing',
    'PaymentMethod': 'payment_method',
    'MonthlyCharges': 'monthly_charges',
    'TotalCharges': 'total_charges',
    'Churn' : 'churn'
}

data.rename(columns=rename_dict, inplace=True)

In [56]:
data.head()

,customer_id,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,online_security,...,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges,churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.50,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [57]:
# Check the number of unique values
unique_counts = data['customer_id'].nunique()
total_rows = len(data)

print(f"Unique Customer IDs: {unique_counts}")
print(f"Total Rows: {total_rows}")

Unique Customer IDs: 7043
Total Rows: 7043


Since the number of unique customer id's match the total number of rows, customer_id is just an identifier and can be dropped 

In [58]:
data.drop(columns=['customer_id'], inplace=True)

In [59]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   gender             7043 non-null   object 
 1   senior_citizen     7043 non-null   int64  
 2   partner            7043 non-null   object 
 3   dependents         7043 non-null   object 
 4   tenure             7043 non-null   int64  
 5   phone_service      7043 non-null   object 
 6   multiple_lines     7043 non-null   object 
 7   internet_service   7043 non-null   object 
 8   online_security    7043 non-null   object 
 9   online_backup      7043 non-null   object 
 10  device_protection  7043 non-null   object 
 11  tech_support       7043 non-null   object 
 12  streaming_tv       7043 non-null   object 
 13  streaming_movies   7043 non-null   object 
 14  contract           7043 non-null   object 
 15  paperless_billing  7043 non-null   object 
 16  payment_method     7043 

In [60]:
cols_to_check = ['online_security', 'online_backup', 'device_protection', 
                 'tech_support', 'streaming_tv', 'streaming_movies', 'multiple_lines']

print("Unique Values")
for col in cols_to_check:
    print(f"{col}: {data[col].unique()}")

Unique Values
online_security: ['No' 'Yes' 'No internet service']
online_backup: ['Yes' 'No' 'No internet service']
device_protection: ['No' 'Yes' 'No internet service']
tech_support: ['No' 'Yes' 'No internet service']
streaming_tv: ['No' 'Yes' 'No internet service']
streaming_movies: ['No' 'Yes' 'No internet service']
multiple_lines: ['No phone service' 'No' 'Yes']


For these columns , 'No internet service' and 'No phone service' is equivalent to 'No' so we convert these to just 'No' for easier analysis

In [61]:
cols_to_fix = ['online_security', 'online_backup', 'device_protection', 
               'tech_support', 'streaming_tv', 'streaming_movies']

for col in cols_to_fix:
    data[col] = data[col].replace('No internet service', 'No')

data['multiple_lines'] = data['multiple_lines'].replace('No phone service', 'No')

In [62]:
all_binary_cols = [
    'partner', 'dependents', 'phone_service', 'multiple_lines', 
    'online_security', 'online_backup', 'device_protection', 
    'tech_support', 'streaming_tv', 'streaming_movies', 
    'paperless_billing', 'churn'
]

for col in all_binary_cols:
    data[col] = data[col].map({'Yes': 1, 'No': 0})

Binary encoded the Yes/No columns for better visualisations and analysis

In [63]:
# checking for constant columns 

data.nunique()

gender                  2
senior_citizen          2
partner                 2
dependents              2
tenure                 73
phone_service           2
multiple_lines          2
internet_service        3
online_security         2
online_backup           2
device_protection       2
tech_support            2
streaming_tv            2
streaming_movies        2
contract                3
paperless_billing       2
payment_method          4
monthly_charges      1585
total_charges        6531
churn                   2
dtype: int64

Since no column has only one unique value, every feature contains at least some variation that could help explain why customers stay or leave, making them all potentially useful for churn analysis.

In [64]:
data.head()

,gender,senior_citizen,partner,dependents,tenure,phone_service,multiple_lines,internet_service,online_security,online_backup,device_protection,tech_support,streaming_tv,streaming_movies,contract,paperless_billing,payment_method,monthly_charges,total_charges,churn
0,Female,0,1,0,1,0,0,DSL,0,1,0,0,0,0,Month-to-month,1,Electronic check,29.85,29.85,0
1,Male,0,0,0,34,1,0,DSL,1,0,1,0,0,0,One year,0,Mailed check,56.95,1889.50,0
2,Male,0,0,0,2,1,0,DSL,1,1,0,0,0,0,Month-to-month,1,Mailed check,53.85,108.15,1
3,Male,0,0,0,45,0,0,DSL,1,0,1,1,0,0,One year,0,Bank transfer (automatic),42.30,1840.75,0
4,Female,0,0,0,2,1,0,Fiber optic,0,0,0,0,0,0,Month-to-month,1,Electronic check,70.70,151.65,1


In [65]:
# save the final cleaned data 

data.to_csv('data_cleaned.csv', index=False)

## DATA ANALYSIS AND EXPLORATION